### DISCUSSION TITLE
Dealing with maintenance windows, running Airflow on Kubernetes (GKE)

### DISCUSSION THREAD

**Discussion by yeachan153** [apache/airflow#24041] [General]
Hi all,

We are currently using Airflow 2.1.4 with celery executor running on Kubernetes (GKE 1.21.11-gke.900).

For GKE we have to specify maintenance windows for node upgrades, and in that time it is not possible to schedule pods on nodes.

Unfortunately, this is problematic as we end up with Airflow tasks failing after the maintenance window finishes and the pods from the Airflow deployment are running again.

We have quite a few hourly DAGs & externalTaskSensor's running in the background throughout the day, so it is not possible to simply avoid having an active dagrun during maintenance windows.

Wondering if others have run into this problem? Is there a plan in the future for airflow to be able to properly recover from maintenance windows?

**Comment by potiuk** [comment]
I wonder - what do you mean by "proper" recovery? What exactly is the problem today? What is the right approach you imagine? Maybe - since you need it - that would be a good idea that you contribute such solution back after discussing it here? 

I think it has never been raised before, but if a good solution for that will be proposed, it might be a good feature for Airlfow.

**Comment by yeachan153** [comment]
Hi @potiuk, thanks for responding :)

Currently when a maintenance window happens, each of the running nodes containing our Airflow deployment in Kubernetes are replaced by some strategy: (in our case with a surge of 1 and max unavailability of 0 - new nodes are added before being removed).

When this occurs, tasks that are currently running in airflow fail. This causes considerable effort on our part to go through all the DAGs we have running (of which there are hundreds) and clear the ones that have failed. We do not want to use some automatic clearing tool because we only want to clear tasks that failed due to some upgrade event/node failure event, not something that failed because some business process within the task failed/timed out. Similarly, we don't want to use the retry option because currently there is no mechanism to retry tasks based on why that failure occurred. Retrying for reasons other than infrastructure failure can be expensive/cause time waste since these will just fail again.

We thought of 2 possible two high level ideas which may work?

1) Kubernetes supports graceful termination by extending a grace period, sending a SIGTERM to the pods within that node to shutdown. We could have the worker use that period to "hand-over" what it is currently processing to another worker during this period somehow? I noticed there was a graceful way of shutting down a worker via `airflow celery stop`, but that waits for the task to complete before shutting down the worker, which would fall outside the grace period. For the scheduler, maybe something similar can be done? In either case we would have to spread all the scheduler pods/worker pods across different nodes.
2) Allow conditional retries based on some log of what caused the failure. This seems easier, but more of a quick fix.

**Reply by potiuk** [reply]
I think retries are far easier to handle. And I think it calls for a feature to set Airflow in a 'maintenance mode' - which would cause any task that failed during the maintenance window to be restarted after the window ends - no matter if it had retry (and should not increase retry counter). I see how it could be useful - you might want to create a feature request for it or - even better - start a PR about it.

**Reply by yeachan153** [reply]
To be honest the first approach feels cleaner from a Kubernetes perspective, but if I understand correctly this approach will be difficult to implement? I would be interested to know what you envision would be challenging for this approach.

Regarding the retries approach with a maintenance mode, I do envision a couple of issues:
- There might still be cases where we retry for genuine failures outside of node failures/upgrades, e.g. business logic fails while maintenance mode is still active
- I originally thought we could just retry for all node shutdowns, not just for node shutdowns caused by node  maintenance/upgrades. I'm not sure if there's much value by distinguishing them? I think we can then just set retries for the task currently executing (and about to fail) because that worker is going away?

It's also unclear if there is a clear notice we can intercept that a maintenance window is starting/ending. At least in GKE we set a maintenance window a few times a week for a few hours - that doesn't necessarily mean all of those hours are maintenance hours - in fact AFAIK far fewer hours are actually used.

**Reply by potiuk** [reply]
I think more important is that Airflow is not a Kuberntes-only - you can use multiple deployments: Celery, local executor, Celery Kubernetes Executor, possibly some other executors in the future. Whatever we implement in Airflow Ideally should not only work for one selected executor (and specific cloud/deployment), but should be possible to use in any of the executors and deployment options. This is why basing it on K8S features and specifically on GKE is not good approach at all.

> There might still be cases where we retry for genuine failures outside of node failures/upgrades, e.g. business logic fails while maintenance mode is still active

I think this is not a big issue and at most it will cause another attempt to do thing that will fail again. Airflow tasks are by definition idempotent so retrying failed tasks should be no issue.  Another option is that we **could** try to distinguish deliberate errors thrown by a task from a forceful shutdown imposed by a node restart. Also you can easily implement failure conditions that should never be retried by throwing AirflowFailException - this is an exception that task should throw if it should not be retried ("hard fail"). In this scenario, we would **not** retry the task if we get AirflowFailException even in maintenance mode as this would indicate "hard" failure rather than one caused by maintenance mode.

> I originally thought we could just retry for all node shutdowns, not just for node shutdowns caused by node maintenance/upgrades. I'm not sure if there's much value by distinguishing them? I think we can then just set retries for the task currently executing (and about to fail) because that worker is going away?

Again - topology of Airflow nodes can be very different for K8S, Celery, Local or even CeleryK8S executor. Also there are scenarios of auto/manual scaliing where number and types of nodes can go up and down. I do not think we should distinguish which nodes are handling which tasks and do different reactions there. Airflow execution model is really 'task" based not "node" based.

> It's also unclear if there is a clear notice we can intercept that a maintenance window is starting/ending. At least in GKE we set a maintenance window a few times a week for a few hours - that doesn't necessarily mean all of those hours are maintenance hours - in fact AFAIK far fewer hours are actually used.

Well I think it's worth checking for GKE case. I believe you can via https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-notifications. But again - the feature should work not only for GKE maintenance mode. This is just one of the many cases, but I think it should handle any kind of maintenance windows:

* manually set maintenance schedule
* signalled by underlying deployment (like cluster notification).
* manually triggered by the user "ad-hoc" via UI (posibly with automated expiration). 
* manually triggered via API (possibly with automated expiration) 

We have to remember, that we are *not* talking about implementing "GKE maintenance windows" - we are talking about generic "maintenance window" feature implemented for Airflow as a product, so it should not be even aware that it is run on GKE. I think there should be a REST API to signal that Airlfow gets into maintenance mode  (possibly with expiration) and possibly some documentation on how to configure GKE cluster notification/what kind of GKE configuration should be done to handle the notifications and convert them in the API calls that would be Airflow-specific, but in general Airflow should not be aware about the actual underlying deployment.

**Reply by yeachan153** [reply]
> I think more important is that Airflow is not a Kuberntes-only - you can use multiple deployments: Celery, local executor, Celery Kubernetes Executor, possibly some other executors in the future. Whatever we implement in Airflow Ideally should not only work for one selected executor (and specific cloud/deployment), but should be possible to use in any of the executors and deployment options. This is why basing it on K8S features and specifically on GKE is not good approach at all.

@potiuk, Sorry, maybe there was too much focus on one specific user case. To be honest this doesn't feel like a Kubernetes only issue. I assume that whatever Airflow process is running (wherever that may be), during some shutdown (caused by whatever, e.g. maintenance) there is a SIGTERM that is received and handled by whatever Airflow process is affected? 

I assume this could be intercepted: workers could intercept that SIGTERM and hand whatever work it's doing back to the scheduler. Something similar could happen for the scheduler: when a SIGTERM gets received it could hand off its scheduling load to another scheduler?

This feels like a more generic/standard way of handling shutdowns rather than creating a new feature in Airflow to handle maintenance windows?

**Reply by potiuk** [reply]
> @potiuk, Sorry, maybe there was too much focus on one specific user case. To be honest this doesn't feel like a Kubernetes only issue. I assume that whatever Airflow process is running (wherever that may be), during some shutdown (caused by whatever, e.g. maintenance) there is a SIGTERM that is received and handled by whatever Airflow process is affected?

> This feels like a more generic/standard way of handling shutdowns rather than creating a new feature in Airflow to handle maintenance windows?

Surely - this is already handled (and has been from as long as I remember). The task can have retries implemented. Ther are many ways task can be stopped an with multiple architecuters tasks can be stopped - and if you have retry set on the task it will be retried (possibly on another node). Again limiting to SIGTERM is to narrow (which by the way is not how it works if you understand how signals work - airflow cannot "capture" SIGTERM sent to  spawned processes - SIGTERM can be sent to such process which will fail and airflow can see the process failing). There are many ways running processes can signal back the failure - it is different in Celery, different in Kubernetes, different in Celery and will be different for future executors.

Currently if such thing happen (and retry is configured for such failing task) airflow will retry it. All fine here. But what I understand you want is to make specific "periods" where such retry will be deferred to AFTER the maintenance window happens. This is what I am talking about. This is what your original request was about right?

**Reply by garywhiteford** [reply]
@potiuk 
What configuration is being described by [this functionality](https://cloud.google.com/composer/docs/composer-2/specify-maintenance-windows#:~:text=An%20Airflow%20worker%20finishes%20already%20started%20tasks%20and%20does%20not%20pick%20new%20tasks.)? Is this part of Apache Airflow proper? 
![image](https://github.com/apache/airflow/assets/33525443/71e26246-b64f-434c-9b5b-3cfac4112671)

**Reply by potiuk** [reply]
I thin you should ask composer team. But likely standard celery feature of putting workers in maintenance mode (See celery documentation).

**Reply by garywhiteford** [reply]
Perfect... thanks!
